pkg/policy/k8s/service.go

```diff
@@ -0,0 +1,278 @@
+	cachedReferences map[*api.K8sServiceNamespace]serviceEndpoints
+}
+
+// emptyMatchedEndpoints returns the empty set
+func emptyMatchedEndpoints() *matchedEndpoints {
+	return &matchedEndpoints{
+		endpoints:        map[k8s.ServiceID]serviceEndpoints{},
+		cachedSelectors:  map[*api.K8sServiceSelectorNamespace]serviceEndpoints{},
+		cachedReferences: map[*api.K8sServiceNamespace]serviceEndpoints{},
+	}
+}
+
+// insertOrGetEndpoints obtains the slice of endpoints for a given service
+func (s *matchedEndpoints) insertOrGetEndpoints(svcID k8s.ServiceID, eps *k8s.Endpoints) serviceEndpoints {
+	res, ok := s.endpoints[svcID]
+	if !ok {
+		res = eps.Prefixes()
+		slices.SortFunc(res, netipx.ComparePrefix)
+		s.endpoints[svcID] = res
+	}
+	return res
+}
+
+// extractMatches stores the endpoints of the passed in service if the service is
+// selected by a ToServices entry in the given rule.
+func (s *matchedEndpoints) extractMatches(rule *api.Rule, svcID k8s.ServiceID, svc *k8s.Service, eps *k8s.Endpoints) {
+	if rule == nil {
+		return
+	}
+	for _, egress := range rule.Egress {
+		for _, toService := range egress.ToServices {
+			if sel := toService.K8sServiceSelector; sel != nil {
+				if serviceSelectorMatches(sel, svcID, svc) {
+					prefixes := s.insertOrGetEndpoints(svcID, eps)
+					s.cachedSelectors[sel] = append(s.cachedSelectors[sel], prefixes...)
+				}
+			} else if ref := toService.K8sService; ref != nil {
+				if serviceRefMatches(ref, svcID) {
+					prefixes := s.insertOrGetEndpoints(svcID, eps)
+					s.cachedReferences[ref] = append(s.cachedReferences[ref], prefixes...)
+				}
+			}
+		}
+	}
+}
+
+// translateRule appends new ToCIDRSet entries for each prefix selected by
+// a ToServices entry. This function requires that extractMatches has been called
+// previously for this rule with each candidate service.
+func (m *matchedEndpoints) translateRule(rule *api.Rule) {
+	if rule == nil {
+		return
+	}
+	for i, egress := range rule.Egress {
+		// Append each prefix as a "Generated" entry, and ensure that we only
+		// append each prefix once (e.g. because multiple ToServices rule select
+		// the same endpoints)
+		appendedPrefix := map[netip.Prefix]struct{}{}
+		appendGeneratedCIDR := func(prefix netip.Prefix) {
+			if _, found := appendedPrefix[prefix]; found {
+				return // avoid inserting same endpoint twice
+			}
+
+			rule.Egress[i].ToCIDRSet = append(rule.Egress[i].ToCIDRSet, api.CIDRRule{
+				Cidr:      api.CIDR(prefix.String()),
+				Generated: true,
+			})
+			appendedPrefix[prefix] = struct{}{}
+		}
+
+		for _, toService := range egress.ToServices {
+			if eps := m.cachedSelectors[toService.K8sServiceSelector]; eps != nil {
+				for _, prefix := range eps {
+					appendGeneratedCIDR(prefix)
+				}
+			} else if eps := m.cachedReferences[toService.K8sService]; eps != nil {
+				for _, prefix := range eps {
+					appendGeneratedCIDR(prefix)
+				}
+			}
+		}
```

Comment:
The Prefix struct seems to be a bit large, I could see this function using up a fair amount of memory. Rather than passing the struct into `appendGeneratedCIDR`, I think it would make sense to only pass the output of the `String()` method. This would save on memory by storing a smaller key in the `appendedPrefix` map (https://go.dev/play/p/ozWzfrAJhea) and by not passing the struct as a value into the function.

```suggestion
		appendedPrefix := map[string]struct{}{}
		appendGeneratedCIDR := func(prefix string) {
			if _, found := appendedPrefix[prefix]; found {
				return // avoid inserting same endpoint twice
			}

			rule.Egress[i].ToCIDRSet = append(rule.Egress[i].ToCIDRSet, api.CIDRRule{
				Cidr:      api.CIDR(prefix),
				Generated: true,
			})
			appendedPrefix[prefix] = struct{}{}
		}

		for _, toService := range egress.ToServices {
			if eps := m.cachedSelectors[toService.K8sServiceSelector]; eps != nil {
				for _, prefix := range eps {
					appendGeneratedCIDR(prefix.String())
				}
			} else if eps := m.cachedReferences[toService.K8sService]; eps != nil {
				for _, prefix := range eps {
					appendGeneratedCIDR(prefix.String())
				}
			}
		}
```

Reply:
Good idea! I didn't consider the fact that we convert to strings anyways, otherwise I would be worried by the additional string allocation (which `netip.Prefix` famously avoids).

Reply:
Marking as resolved since I've removed the dedup logic (see https://github.com/cilium/cilium/pull/31062#issuecomment-1980579585)